hash_writer.go

```diff
@@ -233,41 +235,45 @@ func (hashWriter *hashWriter) Finish() error {
 }
 
 func (hashWriter *hashWriter) Sum(b []byte) ([]byte, error) {
-	return hashWriter.hasher.sum(b)
+	return hashWriter.hasher.Sum(b)
 }
 
 // The following implements HashValue interface.
 
-func (hashWriter *hashWriter) getFieldName() string {
+func (hashWriter *hashWriter) GetFieldName() string {
 	return hashWriter.currentFieldName
 }
 
-func (hashWriter *hashWriter) getAnnotations() []string {
+func (hashWriter *hashWriter) GetAnnotations() []string {
 	return hashWriter.annotations
 }
 
-func (hashWriter *hashWriter) isNull() bool {
+func (hashWriter *hashWriter) IsNull() bool {
 	return hashWriter.currentIsNull
 }
 
-func (hashWriter *hashWriter) ionType() ion.Type {
+func (hashWriter *hashWriter) IonType() ion.Type {
 	return hashWriter.currentType
 }
 
-func (hashWriter *hashWriter) value() (interface{}, error) {
+func (hashWriter *hashWriter) Value() (interface{}, error) {
 	return hashWriter.currentValue, nil
 }
 
-func (hashWriter *hashWriter) isInStruct() bool {
+func (hashWriter *hashWriter) IsInStruct() bool {
 	return hashWriter.currentType == ion.StructType
 }
 
+func (hashWriter *hashWriter) CurrentIsNull() bool {
```

Comment:
Why do we need this new function defined when we already have IsNull()??

Reply:
> Can we move reader_compare.go to internal? We only use it for testing.

Sounds like a good idea. Let's try it.
 
> Also....should we create a testing folder to put all the testing stuff in? (Perhaps that is more a question for @therapon ).

From https://github.com/golang-standards/project-layout it seems like a `test` folder is typically used for **external** tests, or black-box tests, i.e., tests that only rely on public APIs. It would be great if we had these types of tests as well. For white-box (or glass-box) tests I think the pattern is to keep the test files next to the implementation and not export any symbols that client code can depend on.

Reply:
`CurrentIsNull` is enforced by `HashValue` interface, originally it was named as `isNull` but since it is moved to internal package, `isNull` needs to be exportable but IonReader interface already has `IsNull` function so we can't use it

Reply:
Why can't both the HashValue and Ion.Reader interfaces have IsNull()?

Reply:
Because HashReader interface embeds HashValue and Ion.Reader interface and will throw `duplicate method IsNull`

Reply:
The fact that it embeds `IsNull()` from `ion.Reader` doesn't that mean it already has it and we should use it instead of duplicating? 

Let me try and spend sometime in the next couple of days to figure out if we can do without the circularity. It is not clear to me that the circularity is necessary or if it is a side effect of the current code's design.

Reply:
> The fact that it embeds `IsNull()` from `ion.Reader` doesn't that mean it already has it and we should use it instead of duplicating?

The problem is HashWriter which embeds ion.Writer doesn't have `IsNull()`

Reply:
The HashReader/HashWriter interfaces relationship with hashValue was updated in https://github.com/amzn/ion-hash-go/pull/38. Now try to see if we can do without this CurrentIsNull().